Support for SDPA for SAM models

MagnusS0 opened the pull request:
# What does this PR do?

This PR adds support for SDPA to the SAM (Segment Anything Model), following the implementation done by the PyTorch team in [segment-anything-fast](https://github.com/pytorch-labs/segment-anything-fast/).

I add two new classes:

- `SamSdpaAttention(SamAttention)`: For the decoder.
- `SamSdpaVisionAttention(SamVisionAttention)`: For the encoder.

### Performance Improvements (Updated 03.11.2024)
Time Reduction between Eager and SDPA modes for SAM-B and SlimSAM averaged across 10 runs in ms:

| **Batch Size** | **Eager (ms)**      | **SDPA (ms)**       | **Time Reduction (%)** |
|----------------|---------------------|---------------------|------------------------|
| **SAM-B**      |                     |                     |                        |
| 1              | 64.58 (±2.14)       | 30.95 (±0.07)       | 52.06%                 |
| 8              | 1,021.45 (±4.55)    | 227.44 (±0.67)      | 77.73%                 |
| **SlimSAM**    |                     |                     |                        |
| 1              | 50.51 (±1.19)       | 28.70 (±0.03)       | 43.16%                 |
| 8              | 913.97 (±6.73)      | 284.43 (±2.14)      | 68.87%                 |


Peak Memory Usage During Forward Pass:

| **Batch Size** | **Eager (MB)** | **SDPA (MB)** | **Memory Reduction (%)** |
|----------------|----------------|---------------|--------------------------|
| **SAM-B**      |                |               |                          |
| 1              | 2,151          | 875           | 59.31%                   |
| 8              | 16,007         | 3,913         | 75.54%                   |
| **SlimSAM**    |                |               |                          |
| 1              | 1,961          | 1,273         | 35.10%                   |
| 8              | 15,540         | 9,503         | 38.84%                   |


On the TrashNet 1.0 validation set I got a reduction around 30-35%, with no decrease in accuracy.
I also ran training on TrashNet 1.0 with no issues and about 1.7x speedup per epoch with SlimSAM. Also allowed gradient check-pointing that seemed to be missing from the model deceleration, worked without any changes.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@amyeroberts 
@qubvel 
I'll also tag @ArthurZucker since we just worked on this model last weekend.

### Some TODO
Should probably add some documentation for this, but let me know where is best place to add this.

MagnusS0 committed:
feat: add support for sdpa and gradient checkpointing

MagnusS0 committed:
fix: ruff format

MagnusS0 committed:
fix: config sdpa

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_34110). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ArthurZucker commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -856,11 +900,115 @@ def forward(self, hidden_states: torch.Tensor, output_attentions=False) -> torch
+    def __init__(self, config, window_size):
+        super().__init__(config, window_size)
+
+    def add_decomposed_rel_pos(
```
this should be the same as the parrent one no?

ArthurZucker replied:
Oh nice then let's add a small comment about that!

MagnusS0 replied:
Added a small comment about it in d486e627c64c6982ada52805de75857c86ad8c74

ArthurZucker reviewed (commented):
Looks good! 🤗

ArthurZucker reviewed (approved):
Thanks for adding, I think this should be automatically tests, but make sure to run the slow tests

MagnusS0 committed:
fix: sdpa layer naming convention

ArthurZucker commented:
```FAILED tests/models/sam/test_modeling_sam.py::SamModelTest::test_sdpa_can_dispatch_non_composite_models - ValueError: The SDPA model should have SDPA attention layers``` can't merge until it's fixed!

MagnusS0 committed:
fix: update test_eager_matches_sdpa_inference to handle vision_hidden_states

MagnusS0 committed:
test: skip incompatible tests and fix loading issue with sdpa

- Updated tests to skip cases flash and dynamic compile.
- Minor adjustment to ensure correct loading of model with sdpa for dispatch test.

MagnusS0 committed:
style: apply Ruff formatting

MagnusS0 committed:
ruff fix again after rebase

MagnusS0 commented:
Apologies for the commit history mess. I had to rebase to clean things up, so the branch should now be up to date with the main branch and tests pass.

MagnusS0 committed:
Merge branch 'huggingface:main' into sam_support_sdpa

MagnusS0 committed:
[run-slow] sam

MagnusS0 committed:
Merge branch 'main' into sam_support_sdpa

qubvel commented:
Sorry, it should be the last commit, so I can approve the run 🙂

MagnusS0 commented:
> Sorry, it should be the last commit, in that case I can approve the run 🙂

Ahh no problem, will commit again.  Just fyi there are 4 slow tests I expect will fail, as they also do that on main.

MagnusS0 committed:
Merge branch 'main' into sam_support_sdpa

MagnusS0 committed:
[run-slow] sam

qubvel commented:
It looks like everything is fine in CI 🤗 Let's wait for a comment from @zucchini-nlp regarding setting attention implementation in config and I suppose it's ready to be merged

MagnusS0 commented:
Awesome 🤗  Probably just some small differences in my local setup causing the test to fail then.

zucchini-nlp commented on src/transformers/models/sam/configuration_sam.py:
```diff
@@ -299,6 +299,9 @@ def __init__(
         if isinstance(mask_decoder_config, SamMaskDecoderConfig):
             mask_decoder_config = mask_decoder_config.to_dict()
 
+        vision_config["attn_implementation"] = self._attn_implementation
+        mask_decoder_config["attn_implementation"] = self._attn_implementation
+
```
@MagnusS0 hey, no need to do this now, as each sub-config should get its attn implementation set when loading `from_pretrained()`. The to check correctness are `test_attn_implementation_composite_models` and `test_sdpa_can_dispatch_composite_models` and `test_flash_attn_2_can_dispatch_composite_models`  

Also, can you pls add the attributes for `sub_configs` as in this PR (https://github.com/huggingface/transformers/pull/34410/files) and enable tests for configs?

zucchini-nlp commented on tests/models/sam/test_modeling_sam.py:
```diff
@@ -450,6 +451,51 @@ def test_model_from_pretrained(self):
+    def test_sdpa_can_compile_dynamic(self):
+        self.skipTest(reason="SAM model can't be compiled dynamic yet")
+
+    @require_torch_sdpa
+    def test_sdpa_can_dispatch_non_composite_models(self):
+        """
```
it should be the `composite` model as SAM has several sub configs in the main config. You can add class attribute `_is_composite=True` to the tester class so the appropriate tests are called :)

zucchini-nlp replied:
cool, thanks, LGTM! Feel free to ping core maintainer for review :)

MagnusS0 committed:
refactor: Address review comments and improve sub-config handling in SAM model tests

- Added attributes for sub_configs as per PR #34410.
- Enabled tests for configs, ensuring the composite model (SAM) has several sub-configs in the main config.
- Added class attribute _is_composite=True to the tester class
- test_sdpa_can_dispatch_composite_models added

MagnusS0 committed:
[run-slow] sam

MagnusS0 committed:
style: ruff

MagnusS0 committed:
[run-slow] sam

MagnusS0 committed:
style: ruff again ...

MagnusS0 committed:
[run-slow] sam

ArthurZucker reviewed (commented):
Thanks a lot for the addition, sorry for the delay. 
We are going to refactor a bit afterwards in #35235, but let's merge for now! 🤗
